### PR TITLE
Updated Save flow to match the lastest Specification

### PR DESCRIPTION
**Pull request by dxslly**
Migrated save related code to use the new SaveCredentialRequest & SaveCredentialResult messages.

Should resolve: https://github.com/openid/OpenYOLO-Android/issues/33

In a future PR, I can update to use OpenYOLO exceptions being added in PR: https://github.com/openid/OpenYOLO-Android/pull/42

### CODE REVIEW

File: api/java/org/openyolo/api/CredentialClient.java [hunk 1]
```diff
@@ -250,6 +251,32 @@ public HintRetrieveResult getHintRetrieveResult(Intent resultData) {
         }
     }
 
+    public CredentialSaveResult getCredentialSaveResult(Intent resultData) {
```
Comment (iainmcgin) [inline api/java/org/openyolo/api/CredentialClient.java]: `[WARN] /bitrise/src/api/java/org/openyolo/api/CredentialClient.java:254:5: Missing a Javadoc comment. [JavadocMethod]`

File: api/java/org/openyolo/api/CredentialClient.java [hunk 2]
```diff
@@ -137,30 +139,29 @@ public void retrieve(
...
-                findProviders(SAVE_CREDENTIAL_ACTION);
+    public Intent getSaveIntent(final CredentialSaveRequest saveRequest) {
+        List<ComponentName> saveProviders = findProviders(SAVE_CREDENTIAL_ACTION);
 
         if (saveProviders.isEmpty()) {
             return null;
         }
 
-        byte[] encodedCredential = credentialToSave.toProtobuf().toByteArray();
+        final byte[] encodedSaveRequest = saveRequest.toProtocolBuffer().toByteArray();
```
Comment (iainmcgin) [inline api/java/org/openyolo/api/CredentialClient.java]: nit: final is superfluous and inconsistent with the definition of other local variables. I usually only specify final on a local if it is to be used inside a closure / anonymous inner class.

File: api/javatests/org/openyolo/api/CredentialClientTest.java [hunk 3]
```diff
@@ -282,17 +298,57 @@ public void getSaveIntent_threeKnownProviders() throws Exception {
...
                 .endsWith("ProviderPickerActivity");
     }
 
     @Test
+    public void getCredentialSaveResult_withNullIntent_returnsDefaultResult() {
+        final CredentialSaveResult result =
+                credentialClient.getCredentialSaveResult(null /* intent*/);
+
+        assertThat(result.getResultCode()).isEqualTo(CredentialSaveResult.CODE_UNKNOWN);
+        assert(result.getAdditionalProperties()).isEmpty();
```
Comment (iainmcgin) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: assert -> assertThat - I'm actually not sure what the semantics of this is, given that assert is a keyword.

File: api/javatests/org/openyolo/api/CredentialClientTest.java [hunk 4]
```diff
@@ -282,17 +298,57 @@ public void getSaveIntent_threeKnownProviders() throws Exception {
...
+    }
+
+    @Test
+    public void getCredentialSaveResult_missingExtra_returnsDefaultResult() {
+        final Intent intent = new Intent();
+        final CredentialSaveResult result = credentialClient.getCredentialSaveResult(intent);
+
+        assertThat(result.getResultCode()).isEqualTo(CredentialSaveResult.CODE_UNKNOWN);
+        assert(result.getAdditionalProperties()).isEmpty();
```
Comment (iainmcgin) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: assert -> assertThat

File: api/javatests/org/openyolo/api/CredentialClientTest.java [hunk 5]
```diff
@@ -282,17 +298,57 @@ public void getSaveIntent_threeKnownProviders() throws Exception {
...
+    }
+
+    @Test
+    public void getCredentialSaveResult_withInvalidProto_returnsDefaultResult() {
+        final Intent intent = new Intent();
+        intent.putExtra(ProtocolConstants.EXTRA_SAVE_RESULT, INVALID_PROTO);
+        final CredentialSaveResult result = credentialClient.getCredentialSaveResult(intent);
+
+        assertThat(result.getResultCode()).isEqualTo(CredentialSaveResult.CODE_UNKNOWN);
+        assert(result.getAdditionalProperties()).isEmpty();
```
Comment (iainmcgin) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: assert -> assertThat

File: api/javatests/org/openyolo/api/CredentialClientTest.java [hunk 6]
```diff
@@ -282,17 +298,57 @@ public void getSaveIntent_threeKnownProviders() throws Exception {
...
+    }
+
+    @Test
+    public void getCredentialSaveResult_validResult_returnsResult() {
+        final CredentialSaveResult result =
+                credentialClient.getCredentialSaveResult(
+                        CredentialSaveResult.SAVED.toResultDataIntent());
+
+        assertThat(result.getResultCode()).isEqualTo(CredentialSaveResult.CODE_SAVED);
+        assert(result.getAdditionalProperties()).isEmpty();
```
Comment (iainmcgin) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: assert -> assertThat

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    @NonNull
+    private final Map<String, byte[]> mAdditionalProperties;
+
+    /**
+     * Reads a save request from its protocol buffer byte array form.
+     * @throws IOException if the given save request could not be decoded.
+     *
+     * @see #toProtocolBuffer()
+     */
+    public static CredentialSaveRequest fromProtoBytes(byte[] credentialSaveRequestBytes)
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: I have a "fromProto" method in each of the other types; please add for consistency.

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+        return new CredentialSaveRequest.Builder(
+                Protobufs.CredentialSaveRequest.parseFrom(credentialSaveRequestBytes))
+                .build();
+    }
+
+    /**
+     * Creates a {@link CredentialSaveRequest} from the given {@link Credential}.
+     */
+    public static CredentialSaveRequest fromCredential(Credential credential) {
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: Interesting - I should probably add this to the delete request as well.

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+            return this;
+        }
+
+        /**
+         * Specifies additional, non-standard save request parameters. The provided map
+         * must be non-null, and contain only non-empty keys and non-null values.
+         */
+        @NonNull
+        public Builder setAdditionalProperties(@NonNull Map<String, byte[]> additionalParams) {
+            require(additionalParams, notNullValue());
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: AdditionalPropertiesUtil will do these checks for you.

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+     * Parcelable reader for {@link CredentialSaveRequest} instances.
+     * @see android.os.Parcelable
+     */
+    public static final Creator<CredentialSaveRequest> CREATOR = new CredentialSaveRequestCreator();
+
+    @NonNull
+    private final Credential mCredential;
+
+    @NonNull
+    private final Map<String, byte[]> mAdditionalProperties;
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: preferably retain Map<String, ByteString> as then the values are properly immutable.

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    /**
+     * Creates a {@link CredentialSaveRequest} from the given {@link Credential}.
+     */
+    public static CredentialSaveRequest fromCredential(Credential credential) {
+        return new Builder(credential).build();
+    }
+
+    private CredentialSaveRequest(Builder builder) {
+        mCredential = builder.mCredential;
+        mAdditionalProperties = builder.mAdditionalProps;
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: To ensure immutability:
`mAdditionalProperties = Collections.unmodifiableMap(builder.mAdditionalProps)`

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        return mAdditionalProperties;
+    }
+
+    @Override
+    public int describeContents() {
+        return 0;
+    }
+
+    @Override
+    public void writeToParcel(Parcel dest, int flags) {
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: I've been wondering if we should bother making the request / result classes Parcelable - I'll discuss that with you offline.

File: protocol/java/org/openyolo/protocol/CredentialSaveResult.java [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+            return this;
+        }
+
+        /**
+         * Specifies additional, non-standard save result properties. The provided map
+         * must be non-null, and contain only non-empty keys and non-null values.
+         */
+        @NonNull
+        public Builder setAdditionalProperties(@NonNull Map<String, byte[]> additionalProps) {
+            require(additionalProps, notNullValue());
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/CredentialSaveResult.java]: AdditionalPropertiesUtil

File: protocol/java/org/openyolo/protocol/ProtocolConstants.java [hunk 14]
```diff
@@ -50,6 +45,16 @@
...
+     * The extra key value used to carry a credential save request.
+     */
+    public static final String EXTRA_SAVE_REQUEST = "org.openyolo.credential.save.request";
+
+    /**
+     * The extra key value used to carry a credential save result.
+     */
+    public static final String EXTRA_SAVE_RESULT = "org.openyolo.credential.save.result";
+
+    /**
```
Comment (iainmcgin) [inline protocol/java/org/openyolo/protocol/ProtocolConstants.java]: SAVE_CREDENTIAL_ACTION also needs updated to "org.openyolo.credential.save".

Comment (dxslly) [inline protocol/java/org/openyolo/protocol/ProtocolConstants.java]: Done

File: api/java/org/openyolo/api/CredentialClient.java (hunk 2 above)
Comment (dxslly) [inline api/java/org/openyolo/api/CredentialClient.java]: Done

File: api/javatests/org/openyolo/api/CredentialClientTest.java (hunk 3 above)
Comment (dxslly) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: Opps, fixed.

File: api/javatests/org/openyolo/api/CredentialClientTest.java (hunk 4 above)
Comment (dxslly) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: Fixed

File: api/javatests/org/openyolo/api/CredentialClientTest.java (hunk 5 above)
Comment (dxslly) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: Fixed

File: api/javatests/org/openyolo/api/CredentialClientTest.java (hunk 6 above)
Comment (dxslly) [inline api/javatests/org/openyolo/api/CredentialClientTest.java]: Fixed

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java (hunk 7 above)
Comment (dxslly) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: Done

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java (hunk 11 above)
Comment (dxslly) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: +1 Done

File: protocol/java/org/openyolo/protocol/CredentialSaveRequest.java (hunk 9 above)
Comment (dxslly) [inline protocol/java/org/openyolo/protocol/CredentialSaveRequest.java]: Done

File: protocol/java/org/openyolo/protocol/CredentialSaveResult.java (hunk 13 above)
Comment (dxslly) [inline protocol/java/org/openyolo/protocol/CredentialSaveResult.java]: Done

File: api/java/org/openyolo/api/CredentialClient.java (hunk 1 above)
Comment (dxslly) [inline api/java/org/openyolo/api/CredentialClient.java]: Fixed.

**Pull request merged**